Логирование через динамический прокси

prosk opened the pull request:
ДЗ по AOP

prosk committed:
Логирование через динамический прокси

spvgl commented on hw10-byteCodes/src/main/java/ru/otus/aop/proxy/ProxyHelper.java:
```diff
@@ -0,0 +1,59 @@
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            Method implMethod = myClass.getClass().getDeclaredMethod(method.getName(),
+                    method.getParameterTypes());
```
Рефлекшн - штука дорогая.
Каждый раз вычислять методы класса (getDeclaredMethod) при вызове метода может серьезно снизить производительность. 
Можно оптимизировать и сначала заранее вычислить методы, аннотированные `@Log` и запомнить их в какой-то коллекции (кэш). А в `invoke` проверять, что вызываемый метод есть среди запомненных методов.

spvgl commented on hw10-byteCodes/src/main/java/ru/otus/aop/proxy/ProxyHelper.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    static TestLogging createMyClass() {
+        InvocationHandler handler = new DemoInvocationHandler(new TestLoggingImpl());
```
Давайте напишем более универсальную реализацию, чтобы можно было создать прокси для любого класса.

Использовать как-то так (main):
```
TestLogging myTestLogging1 = ProxyHelper.createMyClass(new TestLoggingImpl());
TestLogging myTestLogging2 = ProxyHelper.createMyClass(new TestLoggingOtherImpl());
```

Передаем экземпляр класса в `ProxyHelper` и получаем прокси.
Названия методов, здесь, конечно стоит подправить - оставил ваши, чтобы было понятно.

prosk committed:
1) Добавлена возможность использовать С для любого класса, реализующего интерфейс TestLogging
2) Добавлено предварительное сохранение методов в HashSet в DemoInvocationHandler

prosk commented:
Добавил изменения по замечаниям
